File: leetcode/0536-construct-binary-tree-from-string/solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def str2tree(self, s: str) -> Optional[TreeNode]:
        if not s:
            return None

        self.i = 0

        def parse_int():
            start = self.i
            if s[self.i] == '-':
                self.i += 1
            while self.i < len(s) and s[self.i].isdigit():
                self.i += 1
            return int(s[start:self.i])

        def parse_node():
            val = parse_int()
            node = TreeNode(val)

            if self.i < len(s) and s[self.i] == '(':
                self.i += 1  # consume '('
                node.left = parse_node()
                self.i += 1  # consume ')'

            if self.i < len(s) and s[self.i] == '(':
                self.i += 1  # consume '('
                node.right = parse_node()
                self.i += 1  # consume ')'

            return node

        return parse_node()

File: leetcode/0536-construct-binary-tree-from-string/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private String s;
    private int i;

    public TreeNode str2tree(String s) {
        if (s == null || s.isEmpty()) {
            return null;
        }
        this.s = s;
        this.i = 0;
        return parseNode();
    }

    private int parseInt() {
        int start = i;
        if (s.charAt(i) == '-') {
            i++;
        }
        while (i < s.length() && Character.isDigit(s.charAt(i))) {
            i++;
        }
        return Integer.parseInt(s.substring(start, i));
    }

    private TreeNode parseNode() {
        int val = parseInt();
        TreeNode node = new TreeNode(val);

        if (i < s.length() && s.charAt(i) == '(') {
            i++; // consume '('
            node.left = parseNode();
            i++; // consume ')'
        }

        if (i < s.length() && s.charAt(i) == '(') {
            i++; // consume '('
            node.right = parseNode();
            i++; // consume ')'
        }

        return node;
    }
}
